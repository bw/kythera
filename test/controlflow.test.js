const t = require("./util").test

describe("Control flow", () => {
	describe("if", () => {
		t("if", `
if false {
	a
}	
	`, {compile: false})

		t("if-else", `
if false {
    a
} else {
    b
}	
	`, {compile: false})

		t("if-else-if", `
if false {
    a
} else if false {
    23
} else {
    "yay"
}	
	`, {compile: false})

		t("if-else with boolean", `
if a == b {
    a
} else {
    b
}	
	`, {compile: false})

		t("if-else-if with boolean", `
if a == b {
	a
} else if b == c {
	23
} else {
	"hello"
}	
	`, {compile: false})

	})

	describe("while", () => {
		t("while", `
while a == b {
    a = a + 2
}	
	`, {compile: false})

		t("while", `
let x = 0
while x < 10 {
    x = x + 1
}

x	
	`, {compile: false})
	})
})