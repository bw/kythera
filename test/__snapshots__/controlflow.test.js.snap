// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Types if 1`] = `
Object {
  "body": Array [
    Object {
      "kind": "identifier",
      "name": "a",
    },
  ],
  "condition": ParseNode {
    "kind": "literal",
    "type": "bool",
    "value": false,
  },
  "else": Array [
    Object {
      "kind": "identifier",
      "name": "b",
    },
  ],
  "kind": "if",
}
`;

exports[`Types if-else 1`] = `
Object {
  "body": Array [
    Object {
      "kind": "identifier",
      "name": "a",
    },
  ],
  "condition": ParseNode {
    "kind": "literal",
    "type": "bool",
    "value": false,
  },
  "else": Object {
    "body": Array [
      ParseNode {
        "kind": "literal",
        "type": "int",
        "value": 23,
      },
    ],
    "condition": ParseNode {
      "kind": "literal",
      "type": "bool",
      "value": false,
    },
    "else": Array [
      ParseNode {
        "kind": "literal",
        "type": "str",
        "value": "yay",
      },
    ],
    "kind": "if",
  },
  "kind": "if",
}
`;

exports[`Types if-else with boolean 1`] = `
Object {
  "body": Array [
    Object {
      "kind": "identifier",
      "name": "a",
    },
  ],
  "condition": ParseNode {
    "kind": "binary",
    "left": Object {
      "kind": "identifier",
      "name": "a",
    },
    "operator": "==",
    "right": Object {
      "kind": "identifier",
      "name": "b",
    },
  },
  "else": Object {
    "body": Array [
      ParseNode {
        "kind": "literal",
        "type": "int",
        "value": 23,
      },
    ],
    "condition": ParseNode {
      "kind": "binary",
      "left": Object {
        "kind": "identifier",
        "name": "b",
      },
      "operator": "==",
      "right": Object {
        "kind": "identifier",
        "name": "c",
      },
    },
    "else": Array [
      ParseNode {
        "kind": "literal",
        "type": "str",
        "value": "hello",
      },
    ],
    "kind": "if",
  },
  "kind": "if",
}
`;

exports[`Types if-else-if 1`] = `
Object {
  "body": Array [
    Object {
      "kind": "identifier",
      "name": "a",
    },
  ],
  "condition": ParseNode {
    "kind": "binary",
    "left": Object {
      "kind": "identifier",
      "name": "a",
    },
    "operator": "==",
    "right": Object {
      "kind": "identifier",
      "name": "b",
    },
  },
  "else": Array [
    Object {
      "kind": "identifier",
      "name": "b",
    },
  ],
  "kind": "if",
}
`;

exports[`Types if-else-if with boolean 1`] = `
Object {
  "body": Array [
    ParseNode {
      "kind": "assign",
      "left": Object {
        "kind": "identifier",
        "name": "a",
      },
      "operator": "=",
      "right": ParseNode {
        "kind": "binary",
        "left": Object {
          "kind": "identifier",
          "name": "a",
        },
        "operator": "+",
        "right": ParseNode {
          "kind": "literal",
          "type": "int",
          "value": 2,
        },
      },
    },
  ],
  "condition": ParseNode {
    "kind": "binary",
    "left": Object {
      "kind": "identifier",
      "name": "a",
    },
    "operator": "==",
    "right": Object {
      "kind": "identifier",
      "name": "b",
    },
  },
  "kind": "while",
}
`;

exports[`Types while 1`] = `
Object {
  "identifier": "x",
  "kind": "let",
  "value": ParseNode {
    "kind": "literal",
    "type": "int",
    "value": 0,
  },
}
`;

exports[`Types while 2`] = `
Object {
  "body": Array [
    ParseNode {
      "kind": "assign",
      "left": Object {
        "kind": "identifier",
        "name": "x",
      },
      "operator": "=",
      "right": ParseNode {
        "kind": "binary",
        "left": Object {
          "kind": "identifier",
          "name": "x",
        },
        "operator": "+",
        "right": ParseNode {
          "kind": "literal",
          "type": "int",
          "value": 1,
        },
      },
    },
  ],
  "condition": ParseNode {
    "kind": "binary",
    "left": Object {
      "kind": "identifier",
      "name": "x",
    },
    "operator": "<",
    "right": ParseNode {
      "kind": "literal",
      "type": "int",
      "value": 10,
    },
  },
  "kind": "while",
}
`;
